src/block/Composter.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\block\utils\CompostFactory;
+use pocketmine\data\runtime\RuntimeDataDescriber;
+use pocketmine\item\Item;
+use pocketmine\item\VanillaItems;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use pocketmine\world\particle\CropGrowthEmitterParticle;
+use pocketmine\world\sound\ComposterEmptySound;
+use pocketmine\world\sound\ComposterFillSound;
+use pocketmine\world\sound\ComposterFillSuccessSound;
+use pocketmine\world\sound\ComposterReadySound;
+use function max;
+use function mt_rand;
+
+class Composter extends Transparent{
+
+	public const MIN_LEVEL = 0;
+	public const MAX_LEVEL = 8;
+
+	protected int $fillLevel = self::MIN_LEVEL;
+
+	protected function describeBlockOnlyState(RuntimeDataDescriber $w) : void{
+		$w->boundedInt(4, self::MIN_LEVEL, self::MAX_LEVEL, $this->fillLevel);
+	}
+
+	protected function recalculateCollisionBoxes() : array{
+		$empty_layer = (max(1, 15 - 2 * $this->fillLevel) - (int) ($this->fillLevel === 0)) / 16;
+		$boxes = [AxisAlignedBB::one()->trim(Facing::UP, $empty_layer)];
+
+		foreach (Facing::HORIZONTAL as $side) {
+			$boxes[] = AxisAlignedBB::one()->trim(Facing::opposite($side), 14 / 16);
+		}
+		return $boxes;
+	}
+
+	public function isEmpty() : bool{
+		return $this->fillLevel === self::MIN_LEVEL;
+	}
+
+	public function isReady() : bool{
+		return $this->fillLevel === self::MAX_LEVEL;
+	}
+
+	public function getFillLevel() : int{
+		return $this->fillLevel;
+	}
+
+	/** @return $this */
+	public function setFillLevel(int $fillLevel) : self{
+		if($fillLevel < 0 || $fillLevel > self::MAX_LEVEL){
+			throw new \InvalidArgumentException("Layers must be in range " . self::MIN_LEVEL . " ... " . self::MAX_LEVEL);
+		}
+		$this->fillLevel = $fillLevel;
+		return $this;
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if ($this->fillLevel === 7) {
```

Comment:
I don't think hardcoding this value looks good. A function or constant can be better.

Reply:
> I don't think hardcoding this value looks good. A function or constant can be better.

I couldn't come up with any good names...
const START_CHANGE_LEVEL?

Reply:
> I don't think hardcoding this value looks good. A function or constant can be better.

```
Mojang:
Eventually, the materials will start to break down into a dark, crumbly substance. This is the "mature" stage, and it's full of nutrients that your plants will love! 
```
The final state is `mature` so PRE_MATURE?

Reply:
some other names are `immature`, `intermediate`, `unfinished `, `young compost`...

Reply:
i don't really know, but hardcoded value need to be change for sure.